### ISSUE TITLE
Helm test --logs fails with hook-delete-policy "hook-failed" or "hook-succeed"

### ISSUE THREAD

**Issue by fritzduchardt** [helm/helm#9098]
When I execute:

`helm test <release> --logs`

I get the following error:

`unable to get pod logs for <release>: pods "<release>" not found`

This only happens, if I set `helm.sh/hook-delete-policy` to `hook-failed` or `hook-succeeded`

While in general, it makes sense that no logs can be retrieved, if the test pod was deleted, in Helm2 this used to work. The logs were retrieved before deletion of the pods.

Output of `helm version`: version.BuildInfo{Version:"v3.2.0", GitCommit:"e11b7ce3b12db2941e90399e874513fbd24bcb71", GitTreeState:"clean", GoVersion:"go1.13.10"}

Output of `kubectl version`: Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.3", GitCommit:"1e11e4a2108024935ecfcb2912226cedeafd99df", GitTreeState:"clean", BuildDate:"2020-10-14T12:50:19Z", GoVersion:"go1.15.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18+", GitVersion:"v1.18.9-eks-d1db3c", GitCommit:"d1db3c46e55f95d6a7d3e5578689371318f95ff9", GitTreeState:"clean", BuildDate:"2020-10-20T22:18:07Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}

Cloud Provider/Platform (AKS, GKE, Minikube etc.): EKS

**Comment by bacongobbler** [comment]
Can you confirm this is still an issue with a more recent release of Helm? The most up-to-date release is Helm 3.4.1.

**Comment by hazcod** [comment]
Having the same issue with `% helm version
version.BuildInfo{Version:"v3.4.2", GitCommit:"23dd3af5e19a02d4f4baa5b2f242645a1a3af629", GitTreeState:"dirty", GoVersion:"go1.15.5"}`.

```shell
% helm test --logs --namespace $(NAMESPACE) $(PROJECT)
```

**Comment by wawa0210** [comment]
I also reproduced this issue in the latest version of helm

https://github.com/helm/helm/blob/master/pkg/action/hooks.go#L101-L105

It should be caused by this piece of code. Executing `execHook` and deleting `deleteHookByPolicy` are executed in the same method. This logic may be split ideally, so that some other operations can be inserted in the middle, such as get logs

Give me some time, I will try

/assign

**Comment by giddel** [comment]
The error occurs with helm/v3 v3.4.2  (from inside fluxcd/helm-controller/v0.4.4)
No error occurs from local with helm/v3 v3.5.0

**Comment by laplasz** [comment]
same error with v3.5.3, pod contains following annotations:
```
    "helm.sh/hook": test-success
    "helm.sh/hook-delete-policy": before-hook-creation
```
and the result of the `helm tests --logs`
```
POD LOGS: 000test-start
23:40:49  010test-ip-addresses: pods "010test-ip-addresses" not found
```

**Comment by hazcod** [comment]
Unstale

**Comment by wirwolf** [comment]
I have the same problem. Please fix this issue

**Comment by bacongobbler** [comment]
@wirwolf generally speaking, features and fixes come from community contributions. The core maintainers volunteer their time to maintain the project - we don’t have enough time to provide patches for every bug report. That’s where we need help from the community.

If you’d like to see this fixed, please provide a patch. We look forward to reviewing your code.

**Comment by nhomble** [comment]
I'd be happy to take a look @bacongobbler. Any hunch where to start looking? @wawa0210 pointed to hooks.go but it's unclear where he got with that.

**Comment by krishnendudas1979** [comment]
Any latest update on this issue? 
I am using Helm Version:"v3.6.3". Today, I got the same issue. Unable to get the log details during execution of 
.\helm.exe test <release name> --logs --namespace <namespace lit>

**Comment by hazcod** [comment]
Unstale

**Comment by mfbieber** [comment]
Hi guys, what about my PR, is that of no interest?

**Comment by gtarnaras** [comment]
Still an issue with Version:"v3.7.0"

**Comment by SuperCoolAlan** [comment]
Still an issue with Helm v3.8.1

**Comment by harishbsrinivas** [comment]
Issue persists in v3.9.0


```
➜ mytest git:(staging) ✗ helm test mytest-staging --logs
NAME: mytest-staging
LAST DEPLOYED: Mon May 30 04:43:53 2022
NAMESPACE: default
STATUS: deployed
REVISION: 33
TEST SUITE:     mytest-test-connection
Last Started:   Mon May 30 04:44:02 2022
Last Completed: Mon May 30 04:44:04 2022
Phase:          Succeeded

Error: unable to get pod logs for mytest-test-connection: pods "mytest-test-connection" not found
```

**Comment by sethgupton-mastery** [comment]
😭

**Comment by mfbieber** [comment]
@yxxhero: did you find time to look at the PR? I believe quite some people would benefit from it. Can I do something to help finish that up?

**Comment by krishna-rakesh-dubey-db** [comment]
We can use this "helm.sh/hook-delete-policy" : "before-hook-creation". It will not delete the job and we can check logs.

**Comment by micxer** [comment]
@krishna-rakesh-dubey-db This doesn't help at all. The issue is not being able to fetch the logs before the pod gets cleaned up. Your "solution" just keeps the completed pod lying around.

**Comment by krishna-rakesh-dubey-db** [comment]
@micxer Yes that annotation doesn't deletes the pod. I am using only "helm.sh/hook-delete-policy" : "before-hook-creation" annotation. So only new job details are retained previous one is deleted. Then I use "kubectl logs job.batch/JOB_NAME -c CONTAINER_NAME -n NAMESPACE" command to get the logs of my job.

**Comment by rezasadeghi1** [comment]
People I really recommend you to check the indentations in your HELM charts. I had this problem for a long time and fixed the indentation and it started working again. I'm not saying its the 100% solution but its worth a try.

**Comment by torrescd** [comment]
Hi @joejulian !

There have been any progress?

It should be feasible to print logs before executing the deletion hooks, or to not print logs of deleted pods. See #10566 


I can try to make a PR if you accept help.

Regards!

**Comment by mfbieber** [comment]
Hi, I have submitted a PR a year ago ;) https://github.com/helm/helm/pull/10603
Someone just needs to look into it!

**Comment by torrescd** [comment]
Hi @mfbieber !

Its an interesting feature, I wonder if it can cover my use case. 

We are using a helm chart [dependency  ](https://github.com/localstack/helm-charts/blob/main/charts/localstack/templates/tests/test-connection.yaml) which have a given test the deletion policy.

When I execute the helm test command with --logs, that test is being executed and the logs are failing.

In the cases were we don't have control of our helm test dependencies, but still we want to execute a helm test --logs, what we should do?

Regards!

**Comment by mfbieber** [comment]
Hi @torrescd, the PR should solve the issue. Someone needs to verify and approve it.

For a workaround: I don’t know of any feature how to deactivate the hooks on dependent charts.

Best,
Michelle

**Comment by reitzig** [comment]
As a workaround for a CI pipeline, I exploit knowing the test pod name statically:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: {{ .Release.Name }}-dry-run
  namespace: {{ .Release.Namespace }}
  annotations:
    helm.sh/hook: test
    helm.sh/hook-delete-policy: before-hook-creation
spec:
# snip #
```

So I can clean up manually in the pipeline:

```bash
helm test 'my-release' \
    --namespace 'my-namespace' \
    --logs
kubectl --namespace 'my-namespace' \
    delete pod 'my-release-dry-run'
```
